producer/writer/message_writer.go

```diff
@@ -405,25 +410,23 @@ func (w *messageWriterImpl) SetCutoverNanos(nanos int64) {
 	w.Unlock()
 }
 
-func (w *messageWriterImpl) AddConsumerWriter(addr string, cw consumerWriter) {
+func (w *messageWriterImpl) AddConsumerWriter(cw consumerWriter) {
 	w.Lock()
-	newConsumerWriters := make(map[string]consumerWriter, len(w.consumerWriters)+1)
-	for key, cw := range w.consumerWriters {
-		newConsumerWriters[key] = cw
-	}
-	newConsumerWriters[addr] = cw
+	newConsumerWriters := make([]consumerWriter, 0, len(w.consumerWriters)+1)
```

Comment:
Any reason why we need to allocate a new slice here?

Reply:
This is so that the message writer can lock to just get the list of consumerWriters and perform the writes outside the Lock, see line 291-293 in this file. This requires creating a new list every time we add/remove a consumer writer, but add/remove is rare and this allows the message writer to reduce the time with lock in hot path.